bundles/org.openhab.binding.mqtt.espmilighthub/src/main/java/org/openhab/binding/mqtt/espmilighthub/internal/handler/EspMilightHubHandler.java

```diff
@@ -168,6 +220,117 @@ private void processIncomingState(String messageJSON) {
+        }
+    }
+
+    /**
+     * Scales mireds to 0-100%
+     */
+    private static PercentType scaleMireds(int mireds) {
+        // range in mireds is 153-370
+        // 100 - (mireds - 153) / (370 - 153) * 100
+        if (mireds >= 370) {
+            return PercentType.HUNDRED;
+        } else if (mireds <= 153) {
+            return PercentType.ZERO;
+        }
+        return new PercentType(BIG_DECIMAL_100.subtract(new BigDecimal(mireds).subtract(BIG_DECIMAL_153)
+                .divide(BIG_DECIMAL_217, MathContext.DECIMAL128).multiply(BIG_DECIMAL_100)));
+    }
+
+    private static BigDecimal polynomial_fit(BigDecimal x, BigDecimal[] coefficients) {
+        var result = BigDecimal.ZERO;
+        var x_accumulator = BigDecimal.ONE;
+        // forms K[4]*x^0 + K[3]*x^1 + K[2]*x^2 + K[1]*x^3 + K[0]*x^4
+        // (or reverse the order of terms for the usual way of writing it in academic papers)
+        int i = 0;
+        while (true) {
+            result = result.add(coefficients[coefficients.length - i - 1].multiply(x_accumulator));
+            i++;
+            if (i == coefficients.length)
+                break;
+            x_accumulator = x_accumulator.multiply(x);
+        }
+        return result;
+    }
+
+    // https://www.jkps.or.kr/journal/download_pdf.php?spage=865&volume=41&number=6 (8) and (9)
+    private static HSBType calculateHSBFromColorTemp(int mireds, PercentType brightness) {
+        var cct = BIG_DECIMAL_MILLION.divide(new BigDecimal(mireds), MathContext.DECIMAL128);
+        var cctInt = cct.intValue();
+
+        BigDecimal[] coefficients;
+        // 1667K to 4000K and 4000K to 25000K; no range checks since our mired range fits within this
+        if (cctInt <= 4000) {
+            coefficients = KANG_X_COEFFICIENTS[1];
+        } else {
+            coefficients = KANG_X_COEFFICIENTS[0];
+        }
+        var x = polynomial_fit(BIG_DECIMAL_THOUSAND.divide(cct, MathContext.DECIMAL128), coefficients);
+
+        if (cctInt <= 2222) {
+            coefficients = KANG_Y_COEFFICIENTS[2];
+        } else if (cctInt <= 4000) {
+            coefficients = KANG_Y_COEFFICIENTS[1];
+        } else {
+            coefficients = KANG_Y_COEFFICIENTS[0];
+        }
+        var y = polynomial_fit(x, coefficients);
+        var rawHsb = HSBType.fromXY(x.floatValue() * 100.0f, y.floatValue() * 100.0f);
+        return new HSBType(rawHsb.getHue(), rawHsb.getSaturation(), brightness);
+    }
+
+    // https://www.waveformlighting.com/tech/calculate-color-temperature-cct-from-cie-1931-xy-coordinates/
+    private static int calculateColorTempFromHSB(HSBType hsb) {
+        var xy = hsb.toXY();
+        var x = xy[0].toBigDecimal().divide(BIG_DECIMAL_100);
+        var y = xy[1].toBigDecimal().divide(BIG_DECIMAL_100);
+        var n = x.subtract(BIG_DECIMAL_03320).divide(BIG_DECIMAL_01858.subtract(y), MathContext.DECIMAL128);
+        var CCT_K = polynomial_fit(n, MCCAMY_COEFFICIENTS);
```

Comment:
Please replace var by BigDecimal and rename the variable to respect the naming convetion.